### DISCUSSION TITLE
Workflow for AI Agents

### DISCUSSION THREAD

**Discussion by yzhao244** [serverlessworkflow/specification#915] [General] [answered]
Hi guys

I think it was last year's AWS re:invent,  AWS did demo to showcase some multi-agent collaborations use case and generative AI using Step Functions with agents from Bedrocks. It was really cool actually lol. 
https://www.youtube.com/watch?v=bqCn74FSPyg

I also noticed that langgraph uses concept of state machines to support agentic workflow as well, if I am not mistaken.
https://blog.langchain.dev/langgraph-multi-agent-workflows/
![image](https://github.com/user-attachments/assets/fb1038fe-b660-4724-a6d2-900fdfcddb41)

So I can't stop thinking if possible our SW spec is able to help with defining agentic workflow such as Agent Supervisor, Hierarchical Agent Teams which mentioned in langraph.  I am just interested exploring more use cases with our SW since AI agent has become such hot topics nowdays :)

**Comment by JBBianchi** [comment]
Hi @yzhao244,

Thank you for bringing up this interesting topic!

I have discussed this internally with @cdavernas, but we haven't yet had a proper community discussion about it, especially since the revamp of the DSL.

The infrastructure for executing workflows is indeed a complex issue. Deciding what the DSL should cover and what it should leave out is challenging. In my opinion, it has been intentionally left out because it is more of an implementation concern. However, I am open to further discussion on this matter.

@cdavernas  @fjtirado @matthias-pichler-warrify @ricardozanini wdty ?

**Comment by cdavernas** [comment]
@yzhao244 Thanks for the links, it's extremely interesting, and would IMO definitely apply to SW. 

I however tend to agree with @JBBianchi that these are concerns that would rather be runtime related, as I don't really see a way or a reason to "impose" the Agent Pattern to implementers.

Interestingly enough, however, it is the pattern we have chosen for Synapse, which could DEFINITLY benefit from such (awesome) ideas. We'd be more than happy to explore such features for the upcoming 1.0.0 release of Synapse, which already allows for label-based discrimination of workflow execution by agents (which we actually named `operators`, just like in k8s). 

As said above, and provided you have the time and interest to collaborate on Synapse, it would be awesome to explore those features with you. If so, don't hesitate to reach out on Slack, so that we can plan a call or another!

**Comment by yzhao244** [comment]
@JBBianchi @cdavernas Thanks for quick response. :) .
Yes, I agree explicitly imposing agent pattern may not make sense to implementers.  And actually, after reading the latest SW spec, I found after "Total Refactor" in v1.0-alpha1, the semantics of our SW DSL I personally feel getting close to natural language ways of describing flows.. such as Document, Input, Use, Do, of workflow properties as well as Call, Do, Fork, Emit, For from Task. 

I read another blog describe autonomous agent system which I somehow found our spec may actually do that. lol 
Tools are some services or software functions are Reusable components such as Function, Extensions and so on. 
Planning, Actions are the Do from Task
Short memory is in-context information which can be carries as input/out through Tasks and flow. Long memory may be some external vector DB resource which can be defined in as External Resource. lol (AI sometime has hallucination, hopefully I am not 😄  )

<img width="626" alt="image" src="https://github.com/user-attachments/assets/39d87c19-0117-4f2d-8484-37611e1ce694">
https://lilianweng.github.io/posts/2023-06-23-agent/

**Reply by ricardozanini** [reply]
hey @yzhao244 these agents can't be implemented in containers/subflows and be called by a main flow acting as the supervisor? I mean, like you said from my understanding the spec already has the capabilities to implement this pattern.

**Comment by cdavernas** [comment]
@yzhao244 Would you be so kind to select an answer, if applicable? 
If not, do you have any other question?

**Reply by yzhao244** [reply, marked as answer]
Oh yes. Sorry, I missed your message :) . Btw, I saw this use case from the our Spec and it is pretty cool to illustrate a multi agent use case @cdavernas  
https://github.com/serverlessworkflow/specification/tree/main/use-cases/multi-agent-ai-content-generation
